pkg/pgmodel/metrics/database/database.go

```diff
@@ -0,0 +1,180 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+type Engine interface {
+	Run()
+	IsRunning() bool
+}
+
+type metricsEngineImpl struct {
+	conn      pgxconn.PgxConn
+	ctx       context.Context
+	isRunning atomic.Value
+	metrics   []metricQueryWrap
+}
+
+// NewEngine creates an engine that performs database metrics evaluation every evalInterval.
+// The engine runs predefined queries that returns a BIGINT as the metric value
+// which is then set as a value to Prometheus metric.
+//
+// Note: Make sure to call this only when the database is TimescaleDB. Plain Postgres
+// will cause evaluation errors.
+func NewEngine(ctx context.Context, conn pgxconn.PgxConn) *metricsEngineImpl {
+	engine := &metricsEngineImpl{
+		conn:    conn,
+		ctx:     ctx,
+		metrics: metrics,
+	}
+	engine.isRunning.Store(false)
+	return engine
+}
+
+func (e *metricsEngineImpl) register() {
+	prometheus.MustRegister(getMetrics(e.metrics)...)
+}
+
+func (e *metricsEngineImpl) unregister() {
+	m := getMetrics(e.metrics)
+	for i := range m {
+		prometheus.Unregister(m[i])
+	}
+}
+
+func getMetrics(m []metricQueryWrap) []prometheus.Collector {
+	var metrics []prometheus.Collector
+	for i := range m {
+		metrics = append(metrics, m[i].metric)
+	}
+	return metrics
+}
+
+const (
+	timeout       = time.Minute
+	evalInterval  = time.Minute * 3
+	maxRetryWait  = time.Minute * 30
+	stopThreshold = 5
+)
+
+func (e *metricsEngineImpl) Run() {
+	e.register()
+	e.isRunning.Store(true)
+	go func() {
+		timeoutCount := 0
+		up.Set(1)
+		defer func() {
+			e.isRunning.Store(false)
+			up.Set(0)
+		}()
+		wait := evalInterval
+		for {
+			if err := e.Update(); err != nil {
+				// Consider any error as timeout since we want to limit db executions if updating metrics is erroring out.
+				timeoutCount++
+			} else {
+				// This is to reset the state of engine that was earlier prepared for handling database timeout.
+				// The moment the db is happy with the timeout (when there is no longer a timeout), hence
+				// we resume the normal wait operation and reset the timeout so that previous timeouts do not affect
+				// the next iterations.
+				timeoutCount = 0
+				wait = evalInterval
+			}
+			if timeoutCount > stopThreshold {
+				// Stop the metrics engine temporarily if the batch times out beyond stopThreshold consecutively.
```

Comment:
I don't think we are stopping the engine here. We are just waiting inside of the engine.